### ISSUE TITLE
Cannot install on windows

### ISSUE THREAD

**Issue by pashagolub** [lelit/pglast#7]
Windows 10, Python 3.6.

Full output is:
```
C:\Python36-32\Scripts>python -m pip install pg_query
Collecting pg_query
  Using cached pg_query-0.27.tar.gz
  Ignoring aenum: markers 'python_version < "3.6"' don't match your environment
Requirement already satisfied: setuptools in c:\python36-32\lib\site-packages (from pg_query)
Installing collected packages: pg-query
  Running setup.py install for pg-query ... error
    Complete output from command C:\Python36-32\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\PAVLO~1.GOL\\AppData\\Local\\Temp\\pip-build-8ph4yow2\\pg-query\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\PAVLO~1.GOL\AppData\Local\Temp\pip-m36oza86-record\install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build\lib.win32-3.6
    creating build\lib.win32-3.6\pg_query
    copying pg_query\error.py -> build\lib.win32-3.6\pg_query
    copying pg_query\keywords.py -> build\lib.win32-3.6\pg_query
    copying pg_query\node.py -> build\lib.win32-3.6\pg_query
    copying pg_query\printer.py -> build\lib.win32-3.6\pg_query
    copying pg_query\__init__.py -> build\lib.win32-3.6\pg_query
    copying pg_query\__main__.py -> build\lib.win32-3.6\pg_query
    creating build\lib.win32-3.6\pg_query\enums
    copying pg_query\enums\lockoptions.py -> build\lib.win32-3.6\pg_query\enums
    copying pg_query\enums\nodes.py -> build\lib.win32-3.6\pg_query\enums
    copying pg_query\enums\parsenodes.py -> build\lib.win32-3.6\pg_query\enums
    copying pg_query\enums\pg_class.py -> build\lib.win32-3.6\pg_query\enums
    copying pg_query\enums\primnodes.py -> build\lib.win32-3.6\pg_query\enums
    copying pg_query\enums\__init__.py -> build\lib.win32-3.6\pg_query\enums
    creating build\lib.win32-3.6\pg_query\printers
    copying pg_query\printers\ddl.py -> build\lib.win32-3.6\pg_query\printers
    copying pg_query\printers\dml.py -> build\lib.win32-3.6\pg_query\printers
    copying pg_query\printers\sfuncs.py -> build\lib.win32-3.6\pg_query\printers
    copying pg_query\printers\__init__.py -> build\lib.win32-3.6\pg_query\printers
    running build_ext
    error: [WinError 2] The system cannot find the file specified

    ----------------------------------------
Command "C:\Python36-32\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\PAVLO~1.GOL\\AppData\\Local\\Temp\\pip-build-8ph4yow2\\pg-query\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\PAVLO~1.GOL\AppData\Local\Temp\pip-m36oza86-record\install-record.txt --single-version-externally-managed --compile" failed with error code 1 in C:\Users\PAVLO~1.GOL\AppData\Local\Temp\pip-build-8ph4yow2\pg-query\
```

**Comment by lelit** [comment]
I never tried to build it on Windows machines, and the error is quite obscure, so it's not easy for me to help here. Does the compiler toolchain come with `make`?

I will try to install it on a VM with Win8 (if it still boot...), and report back.

**Comment by lelit** [comment]
A very quick inspection suggests that even the underlying libpg_query library isn't easily compilable on Windows: its Makefile relies on GNU make facilities...

**Comment by pashagolub** [comment]
Oh, I see now. I'll try to build it using MinGW and will let you know

**Comment by lfittl** [comment]
Yes, at this point Windows is unfortunately not supported for libpg_query - the error above probably is caused by a compilation error in the C library code itself.

**Comment by lelit** [comment]
Thank you Lukas! I spent a little on this, but the `nmake` that comes with Visual Studio misses lots of GNU make extensions... 
I don't know if that will require a complete rewrite; also my Makefile uses some gmake features.

**Comment by lfittl** [comment]
Apart from the makefiles the other issue is that the extracted code from Postgres is currently not supporting Windows.

I've created https://github.com/lfittl/libpg_query/issues/44 to track this on the libpg_query repo itself.

**Comment by pashagolub** [comment]
Closing this issue as outdated

**Comment by lelit** [comment]
New year brought [new hope](https://github.com/pganalyze/libpg_query/pull/233) on the subject, sooner or later someone could try it out.

**Comment by lelit** [comment]
I did some experiments, but I ran out of spare time and ideas trying to solve [redefinition errors](https://github.com/lelit/pglast/actions/runs/7505705790/job/20435542016#step:7:358) between PG's `win32_port.h` and MSVC 2022 headers :sob:

**Comment by lelit** [comment]
@lfittl, any idea on what may trigger those hard errors?
Maybe I am including something extra, or in the wrong order?

**Comment by lfittl** [comment]
> @lfittl, any idea on what may trigger those hard errors? Maybe I am including something extra, or in the wrong order?

I think that can happen when the windows system headers get loaded before pg_query gets loaded and is able to set some defines. Specifically, I think the Windows definition of S_IFLNK might get pulled in because WIN32_LEAN_AND_MEAN isn't set the first time windows.h gets loaded (see https://github.com/pganalyze/libpg_query/blob/16-latest/src/postgres/include/port/win32_port.h#L47).

It appears that S_IFLNK is the only hard error in that build, it might be worth to change the [relevant code](https://github.com/pganalyze/libpg_query/blob/16-latest/src/postgres/include/port/win32_port.h#L348) to instead look like this:

```
#ifndef S_IFLNK
#define S_IFLNK S_IFCHR
#endif
```

**Comment by lelit** [comment]
Thank you, I will try to reorder the headers inclusion and see what happens.

**Comment by lelit** [comment]
The good news is that on my local development machine I finally got it working, and all tests are green.
OTOH, the CI [is not happy](https://github.com/lelit/pglast/actions/runs/7569967892/job/20614393470#step:7:648), and it will take some time before I can spend more cycles on that fighting with [Windows](https://github.com/lelit/pglast/tree/windows).

**Comment by lelit** [comment]
I just released 6.0.dev2, and the CI should produce Windows wheels... :partying_face:

**Comment by lelit** [comment]
v6 is out, carrying Windows binary wheels!
